net/udprelay/server.go

```diff
@@ -0,0 +1,517 @@
+// otherwise false.
+func (e *serverEndpoint) isBound() bool {
+	return e.handshakeState[0] == disco.BindUDPRelayHandshakeStateAnswerReceived &&
+		e.handshakeState[1] == disco.BindUDPRelayHandshakeStateAnswerReceived
+}
+
+// NewServer constructs a Server listening on 0.0.0.0:'port'. IPv6 is not yet
+// supported. Port may be 0, and what ultimately gets bound is returned as
+// 'boundPort'. Supplied 'addrs' are joined with 'boundPort' and returned as
+// ServerEndpoint.AddrPorts in response to Server.AllocateEndpoint() requests.
+//
+// TODO: IPv6 support
+// TODO: dynamic addrs:port discovery
+func NewServer(port int, addrs []netip.Addr) (s *Server, boundPort int, err error) {
+	s = &Server{
+		disco:               key.NewDisco(),
+		bindLifetime:        defaultBindLifetime,
+		steadyStateLifetime: defaultSteadyStateLifetime,
+		closeCh:             make(chan struct{}),
+		byDisco:             make(map[pairOfDiscoPubKeys]*serverEndpoint),
+		byVNI:               make(map[uint32]*serverEndpoint),
+	}
+	s.discoPublic = s.disco.Public()
+	// TODO: instead of allocating 10s of MBs for the full pool, allocate
+	// smaller chunks and increase as needed
+	s.vniPool = make([]uint32, 0, 1<<24-1)
+	for i := 1; i < 1<<24; i++ {
+		s.vniPool = append(s.vniPool, uint32(i))
+	}
+	boundPort, err = s.listenOn(port)
+	if err != nil {
+		return nil, 0, err
+	}
+	addrPorts := make([]netip.AddrPort, 0, len(addrs))
+	for _, addr := range addrs {
+		addrPort, err := netip.ParseAddrPort(net.JoinHostPort(addr.String(), strconv.Itoa(boundPort)))
+		if err != nil {
+			return nil, 0, err
+		}
+		addrPorts = append(addrPorts, addrPort)
+	}
+	s.addrPorts = addrPorts
+	s.wg.Add(2)
+	go s.packetReadLoop()
+	go s.endpointGCLoop()
+	return s, boundPort, nil
+}
+
+func (s *Server) listenOn(port int) (int, error) {
+	uc, err := net.ListenUDP("udp4", &net.UDPAddr{Port: port})
+	if err != nil {
+		return 0, err
+	}
+	// TODO: set IP_PKTINFO sockopt
+	_, boundPortStr, err := net.SplitHostPort(uc.LocalAddr().String())
+	if err != nil {
+		s.uc.Close()
+		return 0, err
+	}
+	boundPort, err := strconv.Atoi(boundPortStr)
+	if err != nil {
+		s.uc.Close()
+		return 0, err
+	}
+	s.uc = uc
+	return boundPort, nil
+}
+
+// Close closes the server.
+func (s *Server) Close() error {
+	s.closeOnce.Do(func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.uc.Close()
+		close(s.closeCh)
+		s.wg.Wait()
+		clear(s.byVNI)
+		clear(s.byDisco)
+		s.vniPool = nil
+		s.closed = true
+	})
+	return nil
+}
+
+func (s *Server) endpointGCLoop() {
+	defer s.wg.Done()
+	ticker := time.NewTicker(s.bindLifetime)
+	defer ticker.Stop()
+
+	gc := func() {
+		now := time.Now()
+		// TODO: consider performance implications of scanning all endpoints and
+		// holding s.mu for the duration. Keep it simple (and slow) for now.
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		for k, v := range s.byDisco {
+			if v.isExpired(now, s.bindLifetime, s.steadyStateLifetime) {
```

Comment:
downstream this hits isBound, which checks state fields, so presumably all state transitions are also guarded by/contended by s.mu currently? (edit: yes, I see below).